libraries/rush-lib/src/logic/pnpm/PnpmProjectManifestConfiguration.ts

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import { JsonFile, JsonSchema, FileSystem } from '@rushstack/node-core-library';
+
+/**
+ * Pnpm config in package.json
+ * @beta
+ */
+export interface IPnpmProjectManifestConfigurationJson {
+  /**
+   * Pnpm field in root package.json. See https://pnpm.io/package_json
+   * The type of value is unknown because these fields are not validated by Rush.js
+   */
+  pnpmFieldInRootPackageJson: unknown;
```

Comment:
This property name is a little confusing. What about something like `pnpmExtensions`?

@octogonz - you're good at naming things.

Reply:
> Can we just put these options in a separate file, whose schema is not tracked by Rush (so that new options added to pnpm don't require updates to Rush to support), and whose contents just get directly transcribed into `common/temp/package.json` during the installation process?

👆 This is the part that I was concerned about.  Let me explain:

Rush now delegates installation management to PNPM as much as possible. And we also try to preserve the conventional PNPM behavior as much as possible.  But `rush install` does still wrap PNPM, and for good reasons. Simply passing through any arbitrary PNPM setting has some problems:

1. **Compatibility:** Arbitrary settings may be incompatible with Rush's operation.  There isn't an easy way for Rush to warn about incompatible settings if we're simply passing along a property bag.

2. **Validation:** A wildcard `pnpm-config.schema.json` cannot provide useful validation or IntelliSense tips.

3. **Traps for the unwary:** Some of PNPM's settings are a very bad idea to use in a large monorepo, however the PNPM docs do not convey that very well. In fact, sometimes the docs encourage bad settings because PNPM's primary audience is smaller repos. The monorepo I'm working on right now got burned by this, and we have top engineers. (Using package managers correctly is not easy!) Examples of bad settings:  `ignore-compatibility-db=false` , `shared-workspace-lockfile=false`, `shamefully-hoist=true`, `auto-install-peers=true`

4. **No single source of truth:** Certain settings are already specified elsewhere in a Rush repo. If a setting can be specified in two different places, it is highly confusing if the values are contradictory. Examples: `strict-peer-dependencies=true` also specified via **rush.json**, `link-workspace-packages=false` hardwired by Rush

If someone wants to hack up their PNPM configuration with arbitrary settings, this can already be done using the unvalidated wild west `common/config/rush/.npmrc` file.  Ideally we'd eventually have a comment in that `.npmrc` template saying _"You can override any setting here, and FYI if you do that, you're probably using Rush incorrectly."_  This would not block people like @dmichon-msft from experimenting with new features.

Why are we making `rush/config/pnpm-config.json` at all?  Here's some benefits:

- **A recommended battle-tested model.**  The Rush maintainers started with a small repo, and then spent years of expensive trial+error to arrive at a PNPM strategy that works well at scale.  A lot of these lessons are counterintuitive or grounded in highly arcane technical concepts. Rather than exposing settings so that other companies can go through this same painful learning process, why not encourage them to simply copy the right answer? Do we really need more than one "right answer" for how to install NPM packages?  Different companies do face different requirements, but large monorepos tend to accumulate **_every_** requirement -- which makes us all the same. This is the premise of Rush Stack.

- **Rush-specific documentation.**  Our settings in `pnpm-config.json` can be documented from the perspective of Rush and its niche of large monorepos.

- **Clear designation of what's supported.** If a setting is missing from `pnpm-config.json`, the process of requesting it to be added gives the maintainers a chance to think about how it works with Rush and provide some guidance in the docs. If a setting doesn't appear in this file, that makes it clear that it's usage is not supported.

Reply:
> This property name is a little confusing. What about something like `pnpmExtensions`?

We don't need a setting for `pnpmExtensions` or `pnpmFieldInRootPackageJson`, because their members should correspond to top-level settings in **pnpm-config.json** file.

I'm okay with making it easy to experiment with new PNPM settings, but our configuration needs to clearly communicate that such settings are ***unsupported***:  If you want the Rush maintainers to support a PNPM setting, it must first be exposed via **pnpm-config.json**.

Reply:
🤔 A newcomer would probably be confused to find that `"pnpmOptions"` are specified in **rush.json**, whereas these other settings are specified in **pnpm-config.json**.

Rather than waiting for Rush 6, perhaps what we could do is allow `"pnpmOptions"` in **rush.json** only if **pnpm-config.json** does not exist.  If you add that file, then you must delete `"pnpmOptions"` and move your settings into **pnpm-config.json**.

(It's also slightly odd that `"pnpmVersion"` is specified in **rush.json**, but maybe it's handy to have Rush/Node/PNPM versions together in one file.)

Reply:
Here's a design proposal to solve these concerns: https://github.com/chengcyber/rushstack/pull/5

The unvalidated `"unsupportedPackageJsonSettings"` field will hopefully address @dmichon-msft's requirements 😉

@chengcyber  If we accept this proposal, it implies some extra work:

- Update `pnpm-config.schema.json` to validate all the fields including the nesting structures of fields like `"globalPeerDependencyRules"`
- Reject `"pnpmOptions"` in **rush.json** if the file **pnpm-config.json** exists.  The error message can be like `ERROR: Because the new config file "common/config/rush/pnpm-config.json" is being used, you must remove the old setting "pnpmOptions" from rush.json`
- Implement the extra fields
- Ensure that `unsupportedPackageJsonSettings` takes precedence over Rush's settings of these fields

Out of scope for this PR:
- Map other settings from [PNPM .npmrc](https://pnpm.io/npmrc)  -- ultimately I'd like for `.npmrc` to have the same "USE AT YOUR OWN RISK" disclaimer, but we can pursue that later because it's already technically possible to specify those settings
- I removed `onlyBuiltDependencies` because I don't think it's a good idea for people to use this setting globally

Reply:
@octogonz's proposal looks good to me. If @chengcyber doesn't have time to integrate these changes, I'd be happy to help.